// fyers-api-v3/apiService/apiService.js
"use strict";
let { Config } = require("../config/config");
let errorHandler = require("../errorHandler/errorHandler");
const axios = require("axios");
const https = require('https');
const crypto = require('crypto');
let Logger = require("../logger/log.js")

// Create a custom axios instance with keepAlive for connection reuse
const httpsAgent = new https.Agent({
    keepAlive: true
});

const axiosInstance = axios.create({
    httpsAgent: httpsAgent
});

/**
 * Class to access Fyers API.
 * @class
 */
class FyersApi {
    /**
     * Create a new DataSocket object to acess symboldata websocket.
     * @constructor
     * @param {Object} params - The parameters for initializing the DataSocket.
     * @param {string} params.AccessToken - The access token for authentication.
     * @param {string} params.AppID - AppID of user.
     * @param {string} params.RedirectURL - Redirect URL provided while creating the APP.
     * @param {boolean} params.enableLogging - Flag to enable or disable logging.
     * @param {string} [params.Version="2.0"] - the version of API you want your output format in.
     */
    constructor(params) {
        var self = this;

        /**
         * AccessToken of user.
         * @type {String}
         */
        self.AccessToken = params?.AccessToken ?? null;;

        /**
         * RedirectURL of provided by user.
         * @type {String}
         */
        self.RedirectURL = params?.RedirectURL ?? null;

        /**
         * APPID provided by user.
         * @type {String}
         */
        self.AppID = params?.AppID ?? null;
        /**
         * API version provided by client.
         * @type {String}
         */
        self.Version = params?.Version ?? "2.0";

        /**
         * filepath where are logs are to be saved.
         * @type {string}
         */
        self.LogPath = params?.path ?? undefined;

         /**
         * Flag to enable or disable logging.
         * @type {boolean}
         */
        self.LoggingFlag  = params?.enableLogging ?? true;
        
        /**
         * object used to write logs.
         * @type {Logger}
         */
        self.Logger = new Logger(this.LogPath,this.LoggingFlag)

        /**
         * axios instance for making HTTP requests (shared across all instances).
         * @type {Object}
         */
        self.session = axiosInstance;

    }

    /**
     * sets the APPID.
     * @param {string} req - APPID in format xxxxxx-xxx.
     */
    setAppId = function (req) {
        this.AppID = req
    }

    /**
     * sets the RedirectURL.
     * @param {string} req - pass redirect URL set during creation of APP.
     */
    setRedirectUrl = function (req) {
        this.RedirectURL = req
    }

    /**
     * sets the AccessToken.
     * @param {string} req - pass Access token.
     */
    setAccessToken = function (req) {
        if (this.AppID == null) {
            console.log("Please set APPID using setAppId function")
        }
        else {
            this.AccessToken = this.AppID + ":" + req
        }
    }

    /**
     * get generate-authcode URL to login and generate authcode.
     * @param {Object} req - defines for what the callback function is.
     * @param {string} req.client_id - AppID of user.
     * @param {string} req.redirect_uri - RedirectURL provided by user.
     * @param {string} [req.state] - state value.
     * @returns {string} generate authcode url
     */
    generateAuthCode = function (req) {
        var funcname = 'generateAuthCode'
        var logger = this.Logger
        const client_id = (req && req.client_id) || this.AppID;
        const redirect_uri = (req && req.redirect_uri) || this.RedirectURL;
        const state = (req && req.state) || "sample_state";
        logger.debug('generate authcode response', `${Config.API}generate-authcode?client_id=${client_id}&redirect_uri=${redirect_uri}&response_type=code&state=${state}`, funcname);
        return `${Config.SYNC_API}/generate-authcode?client_id=${client_id}&redirect_uri=${redirect_uri}&response_type=code&state=${state}`
    }

    /**
     * generate access token to use the API's.
     * @param {Object} req - defines for what the callback function is.
     * @param {string} req.client_id - AppID of user.
     * @param {string} req.secret_key - App Secret of user.
     * @param {string} req.auth_code - authcode generated after use of generateAuthCode and login.
     * @param {string} [req.code_verifier] - code verifier.
     * @returns {Object} Object with access token and refresh token if successful
     */
    generate_access_token = async (req) => {
        let session = this.session;
        var funcname = 'generateAuthCode'
        var logger = this.Logger
        const code_verifier = req.code_verifier || "";
        const auth_code = req.auth_code || "";
        const client_id = req.client_id || this.AppID;
        const secret_key = req.secret_key || "";
        const sha256 = getSHA256Hash(`${client_id}:${secret_key}`)
        logger.debug("generate_access_token inputs", { "code_verifier": code_verifier, "auth_code": auth_code, "client_id": client_id, "secret_key": secret_key, "sha256": sha256 }, funcname)
        try {

            if (code_verifier === "") {

                const access_token = await session.post(`${Config.SYNC_API}/validate-authcode`, {
                    grant_type: "authorization_code",
                    code: req.auth_code,
                    appIdHash: sha256

                })
                return access_token.data;
            }
            else {

                const access_token = await session.post(`${Config.API}/validate-authcode`, {
                    grant_type: "authorization_code",
                    code_verifier: req.code_verifier,
                    code: req.auth_code,


                })
                return access_token.data;
            }

        }
        catch (e) {
            var err = new errorHandler(e).getError()
            logger.error("error generating access token", err, funcname)
            return err
        }
    }

    /**
     * get profile data.
     * @returns {Promise} returns profile data on resolve or error message and error code on failure
     */
    get_profile = async (req) => {
        let session = this.session;
        var funcname = 'get_profile'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["get_profile"]
                await session.get((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, err, funcname)
                reject(err)
            }
        })
    }

    /**
     * get funds detail.
     * @returns {Promise} returns funds detail on resolve or error message and error code on failure
     */
    get_funds = async (req) => {
        let session = this.session;
        var funcname = 'get_funds'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["funds"]
                await session.get((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, err, funcname)
                reject(err)
            }
        })
    }

    /**
     * get holdings detail.
     * @returns {Promise} returns holdings detail on resolve or error message and error code on failure
     */
    get_holdings = async (req) => {
        let session = this.session;
        var funcname = 'get_holdings'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["holdings"]
                await session.get((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, err, funcname)
                reject(err)
            }
        })
    }

    /**
     * logout API.
     * @returns {Promise} This invalidates the access token
     */
    logout_user = async (req) => {
        let session = this.session;
        var funcname = 'logout_user'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["logout"]
                await session.post((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })

            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * get orderbook to get order details of order placed on current day.
     * @returns {Promise} returns orderbook detail on resolve or error message and error code on failure
     */
    get_orders = async (req) => {
        let session = this.session;
        var funcname = 'get_orders'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["orders"]
                await session.get((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    }).catch(error => {
                        logger.error(`${funcname} response`, error, funcname)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, err, funcname)
                reject(err)
            }
        })
    }

    /**
     * get orderbook to get order details of gtt order placed on current day.
     * @returns {Promise} returns orderbook detail on resolve or error message and error code on failure
     */
    get_gtt_orders = async (req) => {
        let session = this.session;
        var funcname = 'get_gtt_orders'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["gtt_orders"]
                await session.get((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, err, funcname)
                reject(err)
            }
        })
    }

    /**
     * get orderbook for a specific order ID.
     * @param {Object} req - defines for what the callback function is.
     * @param {string} req.order_id - orderID for which orderbook is required for.
     * @returns {Promise} returns orderbook detail of passed order ID on resolve or error message and error code on failure
     */
    get_filtered_orders = async (req) => {
        let session = this.session;
        var funcname = 'get_filtered_orders'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["orders"] + `?id=${req.order_id}`
                await session.get((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * get positionbook to get position details.
     * @returns {Promise} returns positionbook detail on resolve or error message and error code on failure
     */
    get_positions = async (req) => {
        let session = this.session;
        var funcname = 'get_positions'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["positions"]
                await session.get((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, err, funcname)
                reject(err)
            }
        })
    }

    /**
     * get tradebook to get trade details.
     * @returns {Promise} returns tradebook detail on resolve or error message and error code on failure
     */
    get_tradebook = async (req) => {
        let session = this.session;
        var funcname = 'get_tradebook'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["tradebook"]
                await session.get((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, err, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to place order.
     * @param {Object} req - request data.
     * @returns {Promise} returns orderID of placed order on resolve or error message and error code on failure
     */
    place_order = async (req) => {
        let session = this.session;
        var funcname = 'place_order'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["orders_sync"]
                await session.post((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })

            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to place gtt order.
     * @param {Object} req - request data.
     * @returns {Promise} returns orderID of placed order on resolve or error message and error code on failure
     */
    place_gtt_order = async (req) => {
        let session = this.session;
        var funcname = 'place_gtt_order'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["gtt_orders_sync"]
                await session.post((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                    }
                }).then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
                }
                catch (e) {
                    var err = new errorHandler(e).getError()
                    logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * used to place multileg order.
     * @param {Object} req - request data.
     * @returns {Promise} returns orderID of placed order on resolve or error message and error code on failure
     */
    place_multileg_order = async (req) => {
        let session = this.session;
        var funcname = 'place_multileg_order'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["multileg_orders_sync"]
                await session.post((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })

            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to place multiple order.
     * @param {Object[]} req - request data.
     * @returns {Promise} returns orderID of all placed and if they are failed or successful on resolve or error message and error code on failure
     */
    place_multi_order = async (req) => {
        let session = this.session;
        var funcname = 'place_multi_order'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["multi_orders"]
                await session.post((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to modify placed orders which are not traded yet.
     * @param {Object} req - request data.
     * @returns {Promise} returns orderID of passed order ID and if modification was successful or not resolve or error message and error code on failure
     */
    modify_order = async (req) => {
        let session = this.session;
        var funcname = 'modify_order'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["orders_sync"]
                await session.patch((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to modify placed gtt orders which are not traded yet.
     * @param {Object} req - request data.
     * @returns {Promise} returns orderID of passed order ID and if modification was successful or not resolve or error message and error code on failure
     */
    gtt_modify_order = async (req) => {
        let session = this.session;
        var funcname = 'gtt_modify_order'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["gtt_orders_sync"]
                await session.patch((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to modify multiple placed orders which are not traded yet.
     * @param {Object[]} req - request data.
     * @returns {Promise} returns orderID of passed order ID and if modification was successful or not on resolve or error message and error code on failure
     */
    modify_multi_order = async (req) => {
        let session = this.session;
        var funcname = 'modify_multi_order'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["multi_orders"]
                await session.patch((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to cancel placed orders which are not traded yet.
     * @param {Object} req - request data.
     * @returns {Promise} returns orderID of passed order ID and if cancellation was successful or not on resolve or error message and error code on failure
     */
    cancel_order = async (req) => {
        let session = this.session;
        var funcname = 'cancel_order'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["orders_sync"]
                await session.delete((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    },
                    data:req
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to cancel placed gtt orders which are not traded yet.
     * @param {Object} req - request data.
     * @returns {Promise} returns orderID of passed order ID and if cancellation was successful or not on resolve or error message and error code on failure
     */
    gtt_cancel_order = async (req) => {
        let session = this.session;
        var funcname = 'gtt_cancel_order'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["gtt_orders_sync"]
                await session.delete((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    },
                    data:req
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to cancel multiple placed orders which are not traded yet.
     * @param {Object[]} req - request data.
     * @returns {Promise} returns orderID of passed order ID and if cancellation was successful or not on resolve or error message and error code on failure
     */
    cancel_multi_order = async (req) => {
        let session = this.session;
        var funcname = 'cancel_multi_order'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["multi_orders"]
                await session.delete((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    },
                    data:req
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to exit all positions.
     * @param {Object} req - request data.
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    exit_position = async (req) => {
        let session = this.session;
        var funcname = 'exit_position'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["positions"]
                await session.delete((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    },
                    data: req
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to convert open positions.
     * @param {Object} req - request data.
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    convert_position = async (req) => {
        let session = this.session;
        var funcname = 'convert_position'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["positions"]
                await session.post((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to get market status.
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    market_status = async (req) => {
        let session = this.session;
        var funcname = 'market_status'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.data_Api1 + Config["marketStatus"]
                const profile = await session.get((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, err, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to get historic data of symbol.
     * @param {Object} req - request data.
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    getHistory = async (req) => {
        let session = this.session;
        var funcname = 'getHistory'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.data_Api1 + Config["history"]
                url = generateUrl(url, req)
                await session.get((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to get quotes data of symbols.
     * @param {Array} req - array of symbols.
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    getQuotes = async (req) => {
        let session = this.session;
        var funcname = 'getQuotes'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.data_Api1 + Config["quotes"]
                var symbolstring = ""
                req.forEach(function (element) {
                    symbolstring = symbolstring + element + ','
                })
                symbolstring = symbolstring.slice(0, -1)//remove last ','
                var data = { "symbols": symbolstring }
                const urlParams = new URLSearchParams(data);
                url = url + '?' + urlParams
                await session.get((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to get market depth data of symbols.
     * @param {Array} req - array of symbols.
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    getMarketDepth = async (req) => {
        let session = this.session;
        var funcname = 'getQuotes'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.data_Api1 + Config["market_depth"]
                var symbolstring = ""
                req.symbol.forEach(function (element) {
                    symbolstring = symbolstring + element + ','
                })
                symbolstring = symbolstring.slice(0, -1)//remove last ','
                var data = { "symbol": symbolstring, "ohlcv_flag": req.ohlcv_flag }
                const urlParams = new URLSearchParams(data);
                url = url + '?' + urlParams
                await session.get((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to get option chain data of symbols.
     * @param {Object} req - symbol is mandatory, timestamp and strikecount is optional with default value of current expiry and 50 respectively
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    getOptionChain = async (req) => {
        let session = this.session;
        var funcname = 'getOptionChain'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.data_Api1 + Config["optionChain"]
                url = generateUrl(url, req)
                await session.get((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to get price alert data.
     * @param {Object} [req] - optional request parameters for price alert (query parameters)
     * @param {string} [req.archive] - Set to "1" to get archived alerts, default is "0"
     * @returns {Promise<Object>} returns price alert data on resolve or error message and error code on failure
     * Response structure: {code: 200, message: "", s: "ok", data: {[alertId]: {fyToken: string, symbol: string, alert: {comparisonType, condition, name, type, value, status, createdAt, createdEpoch, ...}}}}
     * Example: {code: 200, message: "", s: "ok", data: {"5682484": {fyToken: "10100000008080", symbol: "NSE:SILVERBEES-EQ", alert: {...}}}}
     */
    getPriceAlert = async (req) => {
        let session = this.session;
        var funcname = 'getPriceAlert'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["priceAlert"]
                if (req && Object.keys(req).length > 0) {
                    url = generateUrl(url, req)
                }
                await session.get((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to create price alert.
     * @param {Object} req - request body for creating price alert
     * @param {string} req.agent - Agent identifier (e.g., "fyers-api")
     * @param {number} req["alert-type"] - Alert type (e.g., 1)
     * @param {string} req.name - Alert name (e.g., "gold alert")
     * @param {string} req.symbol - Symbol (e.g., "NSE:GOLDBEES-EQ")
     * @param {string} req.comparisonType - Comparison type (e.g., "LTP")
     * @param {string} req.condition - Condition (e.g., "GT", "LT", "LTE", "GTE")
     * @param {string} req.value - Alert value (e.g., "99999")
     * @param {string} [req.notes] - Optional notes
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    createPriceAlert = async (req) => {
        let session = this.session;
        var funcname = 'createPriceAlert'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["priceAlert"]
                await session.post((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to modify price alert.
     * @param {Object} req - request body for modifying price alert
     * @param {string} req.alertId - Alert ID to modify (e.g., "3871537")
     * @param {string} req.agent - Agent identifier (e.g., "fyers-api")
     * @param {number} req["alert-type"] - Alert type (e.g., 1)
     * @param {string} req.name - Alert name (e.g., "silver bees")
     * @param {string} req.symbol - Symbol (e.g., "NSE:GOLDBEES-EQ")
     * @param {string} req.comparisonType - Comparison type (e.g., "OPEN", "LTP")
     * @param {string} req.condition - Condition (e.g., "GT", "LT", "LTE", "GTE")
     * @param {number} req.value - Alert value (e.g., 299)
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    modifyPriceAlert = async (req) => {
        let session = this.session;
        var funcname = 'modifyPriceAlert'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["priceAlert"]
                await session.put((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to toggle price alert (enable/disable).
     * @param {Object} req - request body for toggling price alert
     * @param {string} req.alertId - Alert ID to toggle (e.g., "3871590")
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    togglePriceAlert = async (req) => {
        let session = this.session;
        var funcname = 'togglePriceAlert'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["togglePriceAlert"]
                await session.put((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to delete price alert.
     * @param {Object} req - request body for deleting price alert
     * @param {string} req.alertId - Alert ID to delete (e.g., "3871519")
     * @param {string} req.agent - Agent identifier (e.g., "fyers-api")
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    deletePriceAlert = async (req) => {
        let session = this.session;
        var funcname = 'deletePriceAlert'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["priceAlert"]
                await session.delete((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    },
                    data: req
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to create smart order limit.
     * @param {Object} req - request body for create smart order limit
     * @param {string} req.symbol - Symbol (e.g., "NSE:SBIN-EQ")
     * @param {number} req.side - Side (1 for Buy, -1 for Sell)
     * @param {number} req.qty - Quantity
     * @param {string} req.productType - Product type (e.g., "CNC", "INTRADAY")
     * @param {number} req.limitPrice - Limit price
     * @param {number} [req.stopPrice] - Stop price (optional)
     * @param {number} req.orderType - Order type
     * @param {number} req.endTime - End time (Unix timestamp)
     * @param {number} req.hpr - High price range
     * @param {number} req.lpr - Low price range
     * @param {number} req.mpp - Maximum price percentage
     * @param {number} req.onExp - On expiry action
     * @param {number} [req.type] - Type (optional)
     * @param {string} [req.validity] - Validity (e.g., "DAY")
     * @param {number} [req.disclosedQty] - Disclosed quantity (optional)
     * @param {boolean} [req.offlineOrder] - Offline order flag (optional)
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    createSmartOrderLimit = async (req) => {
        let session = this.session;
        var funcname = 'createSmartOrderLimit'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["createSmartorderLimit"]
                await session.post((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }


    /**
     * use to create smart order step.
     * @param {Object} req - request body for create smart order step
     * @param {string} req.symbol - Symbol (e.g., "NSE:SBIN-EQ")
     * @param {number} req.side - Side (1 for Buy, -1 for Sell)
     * @param {number} req.qty - Total quantity
     * @param {string} req.productType - Product type (e.g., "CNC", "INTRADAY")
     * @param {number} req.initQty - Initial quantity
     * @param {number} req.avgqty - Average quantity
     * @param {number} req.avgdiff - Average difference
     * @param {number} req.direction - Direction
     * @param {number} req.limitPrice - Limit price
     * @param {number} req.orderType - Order type
     * @param {number} req.startTime - Start time (Unix timestamp)
     * @param {number} req.endTime - End time (Unix timestamp)
     * @param {number} req.hpr - High price range
     * @param {number} req.lpr - Low price range
     * @param {number} req.mpp - Maximum price percentage
     * @param {number} [req.type] - Type (optional)
     * @param {number} [req.stopPrice] - Stop price (optional)
     * @param {string} [req.validity] - Validity (e.g., "DAY") (optional)
     * @param {number} [req.disclosedQty] - Disclosed quantity (optional)
     * @param {boolean} [req.offlineOrder] - Offline order flag (optional)
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    createSmartOrderStep = async (req) => {
        let session = this.session;
        var funcname = 'createSmartOrderStep'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["createSmartorderStep"]
                await session.post((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to create smart order SIP.
     * @param {Object} req - request body for create smart order SIP
     * @param {string} req.symbol - Symbol (e.g., "NSE:SBIN-EQ")
     * @param {number} req.side - Side (1 for Buy, -1 for Sell)
     * @param {number} req.amount - Amount for SIP
     * @param {string} req.productType - Product type (e.g., "CNC", "INTRADAY")
     * @param {number} req.freq - Frequency
     * @param {number} req.sip_day - SIP day
     * @param {boolean} req.imd_start - Immediate start flag
     * @param {number} req.endTime - End time (Unix timestamp)
     * @param {number} req.hpr - High price range
     * @param {number} req.lpr - Low price range
     * @param {number} req.step_up_freq - Step up frequency
     * @param {number} req.step_up_amount - Step up amount
     * @param {number} [req.qty] - Quantity (optional)
     * @param {number} [req.type] - Type (optional)
     * @param {number} [req.limitPrice] - Limit price (optional)
     * @param {number} [req.stopPrice] - Stop price (optional)
     * @param {string} [req.validity] - Validity (e.g., "DAY") (optional)
     * @param {number} [req.disclosedQty] - Disclosed quantity (optional)
     * @param {boolean} [req.offlineOrder] - Offline order flag (optional)
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    createSmartOrderSIP = async (req) => {
        let session = this.session;
        var funcname = 'createSmartOrderSIP'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["createSmartorderSip"]
                await session.post((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to create smart order trail.
     * @param {Object} req - request body for create smart order trail
     * @param {string} req.symbol - Symbol (e.g., "NSE:SBIN-EQ")
     * @param {number} req.side - Side (1 for Buy, -1 for Sell)
     * @param {number} req.qty - Quantity
     * @param {string} req.productType - Product type (e.g., "CNC", "INTRADAY")
     * @param {number} req.stopPrice - Stop price
     * @param {number} req.jump_diff - Jump difference
     * @param {number} req.orderType - Order type
     * @param {number} req.mpp - Maximum price percentage
     * @param {number} [req.type] - Type (optional)
     * @param {number} [req.limitPrice] - Limit price (optional)
     * @param {string} [req.validity] - Validity (e.g., "DAY") (optional)
     * @param {number} [req.disclosedQty] - Disclosed quantity (optional)
     * @param {boolean} [req.offlineOrder] - Offline order flag (optional)
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    createSmartOrderTrail = async (req) => {
        let session = this.session;
        var funcname = 'createSmartOrderTrail'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["createSmartorderTrail"]
                await session.post((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to modify smart order.
     * @param {Object} req - request body for modify smart order
     * @param {string} req.flowId - Flow ID of the smart order to modify
     * @param {number} [req.qty] - New quantity (optional)
     * @param {number} [req.limitPrice] - New limit price (optional)
     * @param {number} [req.endTime] - New end time (Unix timestamp) (optional)
     * @param {number} [req.stopPrice] - New stop price (optional)
     * @param {number} [req.hpr] - New high price range (optional)
     * @param {number} [req.lpr] - New low price range (optional)
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    modifySmartOrder = async (req) => {
        let session = this.session;
        var funcname = 'modifySmartOrder'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["modifySmartorder"]
                await session.patch((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to cancel smart order.
     * @param {Object} req - request body for cancel smart order
     * @param {string} req.flowId - Flow ID of the smart order to cancel
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    cancelSmartOrder = async (req) => {
        let session = this.session;
        var funcname = 'cancelSmartOrder'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["cancelSmartorder"]
                await session.delete((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    },
                    data: req
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to pause smart order.
     * @param {Object} req - request body for pause smart order
     * @param {string} req.flowId - Flow ID of the smart order to pause
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    pauseSmartOrder = async (req) => {
        let session = this.session;
        var funcname = 'pauseSmartOrder'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["pauseSmartorder"]
                await session.patch((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to resume smart order.
     * @param {Object} req - request body for resume smart order
     * @param {string} req.flowId - Flow ID of the smart order to resume
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    resumeSmartOrder = async (req) => {
        let session = this.session;
        var funcname = 'resumeSmartOrder'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["resumeSmartorder"]
                await session.patch((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to get smart order book with filter.
     * @param {Object} [req] - optional request parameters for filtering smart order book
     * @param {string[]} [req.exchange] - Filter by exchange (e.g., ["NSE"])
     * @param {number[]} [req.side] - Filter by side (1 for Buy, -1 for Sell) (e.g., [1, -1])
     * @param {number[]} [req.flowtype] - Filter by flow type (e.g., [1, 2])
     * @param {string[]} [req.product] - Filter by product type (e.g., ["CNC", "INTRADAY"])
     * @param {number[]} [req.messageType] - Filter by message type (e.g., [1, 2])
     * @param {string} [req.search] - Search string (e.g., "SBIN")
     * @param {string} [req.sort_by] - Sort by field (e.g., "CreatedTime")
     * @param {number} [req.ord_by] - Order by (1 for ascending, -1 for descending)
     * @param {number} [req.page_no] - Page number
     * @param {number} [req.page_size] - Page size
     * @returns {Promise} returns smart order book on resolve or error message and error code on failure
     */
    getSmartOrderBookWithFilter = async (req) => {
        let session = this.session;
        var funcname = 'getSmartOrderBookWithFilter'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["smartorderOrderBook"]
                await session.get((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    },
                    params: req
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to create smart exit trigger.
     * @param {Object} req - request body for create smart exit trigger
     * @param {string} req.name - Strategy name (e.g., "Alert Only Strategy")
     * @param {number} req.type - Trigger type (1: Only Alert, 2: Exit with Alert, 3: Exit with Alert + Wait for Recovery)
     * @param {number} req.profitRate - Profit rate threshold
     * @param {number} req.lossRate - Loss rate threshold (negative value)
     * @param {number} [req.waitTime] - Wait time in minutes (required for type 3)
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    createSmartExitTrigger = async (req) => {
        let session = this.session;
        var funcname = 'createSmartExitTrigger'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["smartExitTrigger"]
                await session.post((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to get smart exit trigger.
     * @param {Object} [req] - optional request parameters for getting smart exit trigger
     * @param {string} [req.flowId] - Filter by flow ID (optional)
     * @returns {Promise} returns smart exit trigger details on resolve or error message and error code on failure
     */
    getSmartExitTrigger = async (req) => {
        let session = this.session;
        var funcname = 'getSmartExitTrigger'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["smartExitTrigger"]
                await session.get((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    },
                    params: req
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to update smart exit trigger.
     * @param {Object} req - request body for update smart exit trigger
     * @param {string} req.flowId - Flow ID of the smart exit trigger to update
     * @param {number} [req.profitRate] - New profit rate threshold (optional)
     * @param {number} [req.lossRate] - New loss rate threshold (optional, negative value)
     * @param {number} [req.type] - New trigger type (optional, 1: Only Alert, 2: Exit with Alert, 3: Exit with Alert + Wait for Recovery)
     * @param {string} [req.name] - New strategy name (optional)
     * @param {number} [req.waitTime] - New wait time in minutes (optional, required for type 3)
     * @param {number} [req.triggerPrice] - Trigger price (optional)
     * @param {number} [req.stopLoss] - Stop loss price (optional)
     * @param {number} [req.takeProfit] - Take profit price (optional)
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    updateSmartExitTrigger = async (req) => {
        let session = this.session;
        var funcname = 'updateSmartExitTrigger'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["smartExitTrigger"]
                await session.put((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * use to activate/deactivate smart exit trigger.
     * @param {Object} req - request body for activate/deactivate smart exit trigger
     * @param {string} req.flowId - Flow ID of the smart exit trigger to activate/deactivate
     * @returns {Promise} returns success message on resolve or error message and error code on failure
     */
    activateDeactivateSmartExitTrigger = async (req) => {
        let session = this.session;
        var funcname = 'activateSmartExitTrigger'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["activateSmartExitTrigger"]
                await session.post((url), req, {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        logger.debug(`${funcname} response`, Response.data, funcname)
                        resolve(Response.data)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

     /**
     * get historic orders.
     * @returns {Promise} returns orderbook detail on resolve or error message and error code on failure
     */
    get_order_history = async (req) => {
        let session = this.session;
        var funcname = 'get_order_history'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["orderhistory"]
                if (req && Object.keys(req).length > 0) {
                    url = generateUrl(url, req)
                }
                await session.get((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        resolve(Response.data)
                    }).catch(error => {
                        logger.error(`${funcname} response`, error, funcname)
                        reject(error)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }

    /**
     * get historic trades.
     * @returns {Promise} returns trade history detail on resolve or error message and error code on failure
     */
    get_trade_history = async (req) => {
        let session = this.session;
        var funcname = 'get_trade_history'
        var logger = this.Logger
        let AuthrizationToken = this.AccessToken
        let vers = this.Version
        return new Promise(async function (resolve, reject) {
            try {
                var url = Config.SYNC_API + Config["tradehistory"]
                if (req && Object.keys(req).length > 0) {
                    url = generateUrl(url, req)
                }
                await session.get((url), {
                    headers: {
                        Authorization: AuthrizationToken,
                        version: vers
                    }
                })
                    .then(Response => {
                        resolve(Response.data)
                    }).catch(error => {
                        logger.error(`${funcname} response`, error, funcname)
                        reject(error)
                    })
            }
            catch (e) {
                var err = new errorHandler(e).getError()
                logger.error(`${funcname} response`, { "Error": err, "inputs": req }, funcname)
                reject(err)
            }
        })
    }
    
}
     
function generateUrl(url, queryObject) {
    let finalUrl = url + "?";
    for (let key in queryObject) {
        finalUrl += encodeURIComponent(key) + "=" + encodeURIComponent(queryObject[key]) + "&";
    }
    return finalUrl.slice(0, -1); // Remove the trailing "&"
}

function getSHA256Hash(inputString) {
    const hash = crypto.createHash('sha256');
    hash.update(inputString);
    return hash.digest('hex');
}
module.exports = FyersApi;
module.exports.axiosInstance = axiosInstance;
